tests/utils_tests/test_choices.py

```diff
@@ -2,10 +2,40 @@
 
 from django.db.models import TextChoices
 from django.test import SimpleTestCase
-from django.utils.choices import CallableChoiceIterator, normalize_choices
+from django.utils.choices import (
+    BaseChoiceIterator,
+    CallableChoiceIterator,
+    normalize_choices,
+)
 from django.utils.translation import gettext_lazy as _
 
 
+class SimpleChoiceIterator(BaseChoiceIterator):
+    def __iter__(self):
+        return ((i, f"Item #{i}") for i in range(1, 4))
+
+
+class ChoiceIteratorTests(SimpleTestCase):
+    def test_not_implemented_error_on_missing_iter(self):
+        class InvalidChoiceIterator(BaseChoiceIterator):
+            pass  # Not overriding __iter__().
+
+        msg = "BaseChoiceIterator subclasses must implement __iter__()."
+        with self.assertRaisesMessage(NotImplementedError, msg):
+            iter(InvalidChoiceIterator())
+
+    def test_eq(self):
+        self.assertEqual(
+            SimpleChoiceIterator(),
+            [(1, "Item #1"), (2, "Item #2"), (3, "Item #3")],
+        )
+
```

Comment:
I added this test:

```python
    def test_eq_instances(self):
        self.assertEqual(
            SimpleChoiceIterator(),
            SimpleChoiceIterator(),
        )
```

And it produces this failure:
```
======================================================================
ERROR: test_eq_instances (utils_tests.test_choices.ChoiceIteratorTests.test_eq_instances)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.11/unittest/case.py", line 57, in testPartExecutor
    yield
  File "/usr/lib/python3.11/unittest/case.py", line 623, in run
    self._callTestMethod(testMethod)
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/case.py", line 579, in _callTestMethod
    if method() is not None:
    ^^^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/tests/utils_tests/test_choices.py", line 34, in test_eq_instances
    self.assertEqual(
  File "/usr/lib/python3.11/unittest/case.py", line 873, in assertEqual
    assertion_func(first, second, msg=msg)
    ^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/unittest/case.py", line 863, in _baseAssertEqual
    if not first == second:
    ^^^^^^^^^^^^^^^^^
  File "/home/nessita/fellowship/django/django/utils/choices.py", line 20, in __eq__
    return type(other)(self) == other
    ^^^^^^^^^^^^^^^^^
TypeError: SimpleChoiceIterator() takes no arguments

----------------------------------------------------------------------
```
I think the `__eq__` should work with two instances of the iterator, what do you think?

Reply:
I haven't had to implement much iterables/iterators before, but after some investigation, this seems to be a better implementation for `__eq__`:

```python
return all(a == b for a, b in zip_longest(self, other, fillvalue=object()))
```

@adamchainz @ngnpope what do you think?

Reply:
This might be a bit magical? But I guess relying on the fact that we need to iterate is ok.

Reply:
I guess is a matter of taste, it took me 5 iterations of prints to understand what the original code was doing, hahaha

The fact that this nicely works with any iterable instance, makes me prefer this option.